File: android/app/src/main/java/com/myfabricapp/UgoiraView.java
package com.myfabricapp;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageView;

public class UgoiraView extends ImageView {
    private boolean paused;
    private Bitmap[] frames;
    private int currentFrame;
    private long frameDuration;
    private long lastFrameChangeTime;

    public UgoiraView(Context context) {
        super(context);
        init();
    }

    public UgoiraView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public UgoiraView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        paused = false;
        currentFrame = 0;
        frameDuration = 100; // Default frame duration
    }

    public void setImages(Bitmap[] frames, long frameDuration) {
        this.frames = frames;
        this.frameDuration = frameDuration;
        currentFrame = 0;
        lastFrameChangeTime = System.currentTimeMillis();
        if (!paused) {
            startAnimation();
        }
    }

    private void startAnimation() {
        post(animationRunnable);
    }

    private final Runnable animationRunnable = new Runnable() {
        @Override
        public void run() {
            if (!paused && frames != null && frames.length > 0) {
                long currentTime = System.currentTimeMillis();
                if (currentTime - lastFrameChangeTime >= frameDuration) {
                    currentFrame = (currentFrame + 1) % frames.length;
                    setImageDrawable(new BitmapDrawable(getResources(), frames[currentFrame]));
                    lastFrameChangeTime = currentTime;
                }
                postDelayed(this, frameDuration);
            }
        }
    };

    public void pauseAnimation() {
        paused = true;
    }

    public void resumeAnimation() {
        if (paused) {
            paused = false;
            startAnimation();
        }
    }

    public void setPaused(boolean paused) {
        this.paused = paused;
        if (paused) {
            pauseAnimation();
        } else {
            resumeAnimation();
        }
    }
}

File: android/app/src/main/java/com/myfabricapp/UgoiraView.kt
package com.myfabricapp

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.SurfaceHolder
import android.view.SurfaceView
import kotlin.concurrent.thread

class UgoiraView(context: Context, attrs: AttributeSet) : SurfaceView(context, attrs) {
    private var isPaused: Boolean = false
    private val paint: Paint = Paint()
    private val surfaceHolder: SurfaceHolder = holder
    private var animationThread: Thread? = null

    init {
        setWillNotDraw(false)
    }

    fun setPaused(paused: Boolean) {
        isPaused = paused
    }

    fun pauseAnimation() {
        isPaused = true
        animationThread?.interrupt()
        animationThread = null
    }

    fun resumeAnimation() {
        if (animationThread == null || !animationThread!!.isAlive) {
            isPaused = false
            startAnimation()
        }
    }

    private fun startAnimation() {
        animationThread = thread(start = true) {
            while (!isPaused) {
                try {
                    drawFrame()
                    Thread.sleep(16) // Approx 60 FPS
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                } catch (e: Exception) {
                    // Handle specific exceptions
                }
            }
        }
    }

    private fun drawFrame() {
        val canvas: Canvas? = surfaceHolder.lockCanvas()
        if (canvas != null) {
            try {
                canvas.drawColor(0, android.graphics.PorterDuff.Mode.CLEAR)
                // Draw Ugoira images here
            } finally {
                surfaceHolder.unlockCanvasAndPost(canvas)
            }
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (!isPaused) {
            drawFrame()
        }
    }
}